### PR TITLE
Add NewLazyLexer to defer rules definitions and reduce init costs

### PR DESCRIPTION
**Pull request by moorereason**
Add NewLazyLexer and MustNewLazyLexer which accept a function that
returns the rules for the lexer.  This allows us to defer the rules
definitions until they're needed.

Lexers in a, g, s, and x packages have been updated to use the new lazy
lexer.

Updates #438

### CODE REVIEW

File: regexp.go [hunk 1]
```diff
@@ -422,7 +447,44 @@ restart:
 	return nil
 }
 
+func (r *RegexLexer) compileRules() error {
+	// TODO: refactor to remove this duplication of code from NewLexer
```
Comment (alecthomas) [inline regexp.go:451]: I think `NewLexer()` might be able to just call `return NewLazyLexer(config, func() Rules{ return rules })`?

File: regexp.go [hunk 2]
```diff
@@ -422,7 +447,44 @@ restart:
...
+	}
+
+	r.rules = compiledRules
+	return nil
+}
+
 func (r *RegexLexer) Tokenise(options *TokeniseOptions, text string) (Iterator, error) { // nolint
+	var err error
+	if r.compilerFunc != nil {
+		r.compileOnce.Do(func() {
```
Comment (alecthomas) [inline regexp.go:481]: Nice.

**Review by alecthomas — APPROVED** [review]
I think this change is awesome. If you'd like to finish the rest of the lexers that would be amazing.

File: regexp.go (hunk 1 above)
Comment (moorereason) [reply regexp.go]: Great idea.  Done.

**Comment by moorereason** [comment]
Ok.  This is mostly ready.  I tried to put all the simple lexer conversions into 5972fca (missed the JSON lexer early on).

This gets chroma init costs down to around 2.17ms, 798K, and 3471 allocs for me.

All lexers are updated except for Javascript.  [`lexers/j.Javascript`](https://github.com/alecthomas/chroma/blob/5da831672dbfc21063a28cc79589caf952714382/lexers/j/javascript.go#L9) is public.  I'm assuming you don't want to break compatibility, so I didn't update that lexer.

PS - should update `_tools/pygments2chroma.py`

**Review by alecthomas — APPROVED** [review]

File: regexp.go [hunk 3]
```diff
@@ -188,49 +188,46 @@ func (r Rules) Merge(rules Rules) Rules {
 	return out
 }
 
-// MustNewLexer creates a new Lexer or panics.
-func MustNewLexer(config *Config, rules Rules) *RegexLexer {
-	lexer, err := NewLexer(config, rules)
+// MustNewLazyLexer creates a new Lexer which deferred rules generation.
```
[the remark this answers is not in the document]
Comment (moorereason) [reply regexp.go]: Fixed.

File: regexp.go [hunk 4]
```diff
@@ -188,49 +188,46 @@ func (r Rules) Merge(rules Rules) Rules {
...
 	if err != nil {
 		panic(err)
 	}
 	return lexer
 }
 
-// NewLexer creates a new regex-based Lexer.
+// NewLazyLexer creates a new regex-based Lexer with deferred rules generation.
 //
 // "rules" is a state machine transitition map. Each key is a state. Values are sets of rules
```
[the remark this answers is not in the document]
Comment (moorereason) [reply regexp.go]: I removed the last section of this godoc comment.

**Comment by alecthomas** [comment]
Awesome, thanks!

**Pull request merged**